Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -3057,16 +3057,27 @@ public class CheckHasNoAssumes_Visitor : BottomUpVisitor {
         // to make it more target-language idiomatic and improve performance
         TrStmtList(s.ResolvedStatements, wr);
 
-      } else if (stmt is ExpectStmt) {
-        var s = (ExpectStmt)stmt;
+      } else if (stmt is ExpectStmt || (stmt is AssertStmt && Attributes.Contains(stmt.Attributes, "expect"))) {
```

Comment:
Just an observation: this also could have been written as a Rewriter that literally replaces `assert {:expect} P` with `assert P; expect P;`. The only possible tricky part is ensuring error messages still refer to the original tokens etc. but that doesn't seem too hard.

Not asking for a rewrite (especially since this one is pretty easy to do inline), but do want to make sure we identify things that create friction when pushing compilation towards more independent passes.

Reply:
That is a good idea - and I thought about it, but I'm not as familiar with rewriters. Welcome Rustans' opinion here. The current implementation does have a problem with resolution of the default message that I had to work around. A Rewriter seems conceptually cleaner.

Reply:
More thought: To add in an expect statement, one would have to make a clone of the whole AST I think (correction, one can replace the list elements of a block statement that the assert statement is in).
Also it would alter the AST, which is a problem for the formatter.
And we don't solve the problem of resolving the generated expression that is the default message.
So I'm not sure this is any cleaner.

Reply:
That's just fine and this isn't blocking: I don't actually think it's worth the effort to rewrite this, now that it's implemented and working. But I'm glad I asked because at least we're identifying what speed bumps make it harder to write features as passes.

I've been considering proposing a macro system for Dafny and this seems like a decent use case. :)

Reply:
Regarding the question of whether or not to use a rewriter. It's been my repeated experience that rewriters look appealing at first, but then they cause problems with later passes (including resolution) want to have the original form. For the situation at hand, I suggest not doing a rewrite and instead marking up the AST node (with the attribute, as would naturally happen from parsing, or using a boolean field, if the source-code mechanism were not an attribute).

Reply:
It seems like this feature would be perfect for a post-resolution rewrite. The `"assert with {{:expect}} statement is not allowed in this context (because this is a ghost method or because the statement is guarded by a specification-only expression)"` message would be produced during the rewrite, not after, so there is no issue with producing error messages for an already transformed AST.

If we allowed rewrites that happen right before compilation, after verification, we could even replace `assert {:expect} E` with `expect E` instead of `assert E; expect E`.